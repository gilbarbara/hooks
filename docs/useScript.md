# useScript

Dynamically loads a script by creating a `<script>` tag and appending it to the `document.body`.  
This hook is useful for integrating third-party scripts like APIs or analytics libraries.

## Usage

```tsx
import { useScript } from '@gilbarbara/hooks';

function Component() {
  const [loaded, error] = useScript(
    'https://maps.googleapis.com/maps/api/js?key=Ab2S23Bd&callback=initMap',
    'google-maps'
  );

  return (
    <div>
      {loaded && <div id="maps"/>}
      {error && <span>error</span>}
    </div>
  );
}
```

**Custom Attributes**
  
```tsx
import { useScript } from '@gilbarbara/hooks';

function Component() {
  const [loaded] = useScript('https://example.com/script.js', {
    id: 'example-script',
    async: true,
    type: 'module',
  });

  return <div>{loaded ? 'Script is ready!' : 'Loading...'}</div>;
}
```

## Reference

```typescript
interface UseScriptOptions {
  async?: boolean;
  defer?: boolean;
  id?: string;
  type?: string;
}

type UseScriptResult = [loaded: boolean, error: boolean];

useScript(src: string, idOrOptions: string | UseScriptOptions = {}): UseScriptResult;
```
