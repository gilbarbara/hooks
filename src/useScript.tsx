import { useCallback, useEffect, useMemo, useRef, useState } from 'react';

import { canUseDOM, isString, off, on } from './utils';

interface UseScriptOptions {
  async?: boolean;
  defer?: boolean;
  id?: string;
  type?: string;
}

export type UseScriptResult = [loaded: boolean, error: boolean];

export function useScript(
  src: string,
  idOrOptions: string | UseScriptOptions = {},
): [loaded: boolean, error: boolean] {
  const options = useMemo(
    () => (isString(idOrOptions) ? { id: idOrOptions } : idOrOptions),
    [idOrOptions],
  );
  const script = useRef<HTMLScriptElement>();
  const [state, setState] = useState({
    loaded: false,
    error: false,
  });

  const onLoad = useCallback(() => {
    setState({
      loaded: true,
      error: false,
    });
  }, []);

  const onError = useCallback(() => {
    if (script.current) {
      script.current.remove();
    }

    setState({
      loaded: false,
      error: true,
    });
  }, []);

  useEffect(
    () => {
      if (!canUseDOM() || script.current) {
        return undefined;
      }

      const element = document.createElement('script');

      element.async = options.async ?? true;
      element.defer = options.defer ?? false;
      element.type = options.type || 'text/javascript';
      element.id = options.id || src;
      element.src = src;

      script.current = element;

      const { current } = script;

      on(current, 'load', onLoad);
      on(current, 'error', onError);

      // Add script to document body
      document.body.appendChild(current);

      return () => {
        off(current, 'load', onLoad);
        off(current, 'error', onError);
      };
    },
    [onError, onLoad, options, src], // Only re-run effect if script src changes
  );

  return [state.loaded, state.error];
}
