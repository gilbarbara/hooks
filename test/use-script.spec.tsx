import * as React from 'react';
import { render, waitForElement } from '@testing-library/react';

import useScript from '../src/use-script';

describe('useScript', () => {
  const Component = () => {
    const [loaded] = useScript(
      'https://maps.googleapis.com/maps/api/js?key=Ab2S23Bd&callback=initMap',
    );

    return <div>{loaded && <span>loaded</span>}</div>;
  };

  const { container } = render(<Component />);

  it('should re-render the id has changed', async () => {
    expect(container.firstChild).toMatchSnapshot();

    const script = await waitForElement(() => document.body.querySelector('script'));

    expect(script).toHaveAttribute(
      'src',
      'https://maps.googleapis.com/maps/api/js?key=Ab2S23Bd&callback=initMap',
    );
  });
});
