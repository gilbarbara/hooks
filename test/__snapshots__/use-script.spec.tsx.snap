// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useScript should re-render the id has changed 1`] = `<div />`;
